File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/DruidMeasurementsRepository.java
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments;

import com.google.common.base.Supplier;
import com.google.common.base.Suppliers;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentMeasurements;
import pl.allegro.experiments.chi.chiserver.domain.experiments.MeasurementsRepository;
import pl.allegro.experiments.chi.chiserver.infrastructure.druid.DruidClient;
import pl.allegro.experiments.chi.chiserver.infrastructure.druid.DruidException;

import java.util.*;
import java.util.concurrent.TimeUnit;

public class DruidMeasurementsRepository implements MeasurementsRepository {
    private static final Logger logger = LoggerFactory.getLogger(DruidMeasurementsRepository.class);
    private static final int DRUID_QUERY_TIMEOUT_MS = 2000;
    private static final long CACHE_EXPIRE_DURATION_MINUTES = 30L;

    private final DruidClient druid;
    private final Gson jsonConverter;
    private final String datasource;
    private final Supplier<Map<String, Integer>> lastDayVisitsCache;

    DruidMeasurementsRepository(
            DruidClient druid,
            Gson jsonConverter,
            String datasource) {
        this.druid = druid;
        this.jsonConverter = jsonConverter;
        this.datasource = datasource;
        this.lastDayVisitsCache = Suppliers.memoizeWithExpiration(() -> {
            try {
                return queryLastDayVisits();
            } catch (DruidException e) {
                logger.warn("Error while trying to query Druid for last day visits", e);
                return Collections.emptyMap();
            }
        }, CACHE_EXPIRE_DURATION_MINUTES, TimeUnit.MINUTES);
    }

    private ExperimentMeasurements createExperimentMeasurements(String expId, Map<String, Integer> lastDayVisits) {
        Integer lastDayVisitsNullable = lastDayVisits.get(expId);

        return new ExperimentMeasurements(lastDayVisitsNullable != null ? lastDayVisitsNullable : 0);
    }

    @Override
    public ExperimentMeasurements getMeasurements(String experimentId) {
        return new ExperimentMeasurements(lastDayVisitsCache.get().get(experimentId));
    }

    private Map<String, Integer> queryLastDayVisits() {
        String query = "{\n" +
                "            \"queryType\": \"topN\",\n" +
                "            \"dataSource\": \"" + this.datasource + "\",\n" +
                "            \"intervals\": \"" + DruidClient.lastDayIntervals() + "\",\n" +
                "            \"granularity\": \"all\",\n" +
                "            \"context\": {\n" +
                "            \"timeout\": " + DRUID_QUERY_TIMEOUT_MS +
                "        },\n" +
                "            \"dimension\": {\n" +
                "            \"type\": \"default\",\n" +
                "            \"dimension\": \"experiment_id\",\n" +
                "            \"outputName\": \"experiment_id\"\n" +
                "        },\n" +
                "            \"aggregations\": [\n" +
                "            {\n" +
                "                \"name\": \"sum_visit_count\",\n" +
                "                \"type\": \"doubleSum\",\n" +
                "                \"fieldName\": \"visit_count\"\n" +
                "            }\n" +
                "            ],\n" +
                "            \"metric\": \"sum_visit_count\",\n" +
                "            \"threshold\": 50\n" +
                "        }\n";
        return Optional.of(query)
                .map(it -> druid.query(it))
                .map(it -> jsonConverter.fromJson(it, JsonArray.class))
                .map(it -> it.size() > 0 ? it.get(0) : null)
                .map(it -> {
                    Map<String, Integer> result = new HashMap<>();
                    it.getAsJsonObject()
                            .get("result")
                            .getAsJsonArray()
                            .forEach(e -> result.put(e.getAsJsonObject().get("experiment_id").getAsString(),
                                    e.getAsJsonObject().get("sum_visit_count").getAsInt()));
                    return result;
                })
                .orElse(new HashMap<>());
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/DruidMeasurementsRepository.kt
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments

import com.github.salomonbrys.kotson.*
import com.google.common.base.Suppliers
import com.google.gson.JsonArray
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentMeasurements
import pl.allegro.experiments.chi.chiserver.domain.experiments.MeasurementsRepository
import pl.allegro.experiments.chi.chiserver.infrastructure.JsonConverter
import pl.allegro.experiments.chi.chiserver.infrastructure.druid.DruidClient
import pl.allegro.experiments.chi.chiserver.infrastructure.druid.DruidException
import pl.allegro.experiments.chi.chiserver.infrastructure.druid.lastDayIntervals
import pl.allegro.experiments.chi.chiserver.logger
import java.util.concurrent.TimeUnit

private const val DRUID_QUERY_TIMEOUT_MS = 2000
private const val CACHE_EXPIRE_DURATION_MINUTES = 30L

class DruidMeasurementsRepository(private val druid: DruidClient,
                                  private val jsonConverter: JsonConverter,
                                  private val datasource: String) : MeasurementsRepository {

    private var lastDayVisits = emptyMap<String, Int>()

    companion object {
        private val logger by logger()
    }

    private val lastDayVisitsCache = Suppliers.memoizeWithExpiration(
        {
            try {
                lastDayVisits = queryLastDayVisits()
            } catch (e: DruidException) {
                logger.warn("Error while trying to query Druid for last day visits", e)
            }
            lastDayVisits
        },
        CACHE_EXPIRE_DURATION_MINUTES, TimeUnit.MINUTES)

    private fun asMeasuredExperiment(ex: Experiment, lastDayVisits: Map<String, Int> = this.lastDayVisits) =
        with(ex) {
            Experiment(id, variants, description, documentLink, author, groups, reportingEnabled, activityPeriod,
                ExperimentMeasurements(lastDayVisits[id] ?: 0), editable, null, ex.getExplicitStatus().orElse(null))
        }

    override fun withMeasurements(experiment: Experiment): Experiment = asMeasuredExperiment(experiment)

    override fun withMeasurements(experiments: List<Experiment>): List<Experiment> {
        val lastDayVisits = lastDayVisitsCache.get()
        return experiments.map { asMeasuredExperiment(it, lastDayVisits) }
    }

    private fun queryLastDayVisits(): Map<String, Int> =
        """{
            "queryType": "topN",
            "dataSource": "$datasource",
            "intervals": "${lastDayIntervals()}",
            "granularity": "all",
            "context": {
            "timeout": $DRUID_QUERY_TIMEOUT_MS
        },
            "dimension": {
            "type": "default",
            "dimension": "experiment_id",
            "outputName": "experiment_id"
        },
            "aggregations": [
            {
                "name": "sum_visit_count",
                "type": "doubleSum",
                "fieldName": "visit_count"
            }
            ],
            "metric": "sum_visit_count",
            "threshold": 50
        }
        """.let { druid.query(it) }
            .let { jsonConverter.fromJson<JsonArray>(it) }
            .let { if (it.size() > 0) it[0] else null }
            ?.let {
                it["result"].array.associateBy(
                    { it["experiment_id"].string },
                    { it["sum_visit_count"].int })
            }
            ?: emptyMap()
}
